bin/_test-helpers.sh

```diff
@@ -0,0 +1,318 @@
+#!/bin/bash
+
+# Override CI's `set -e` default, so we can catch errors manually and display
+# proper messages
+set +e
+
+##### Test setup helpers #####
+
+export test_names=(upgrade helm helm-upgrade uninstall deep external-issuer)
+
+handle_input() {
+  export images=""
+  export images_host=""
+  export test_name=""
+  export skip_kind_create=""
+
+  while :
+  do
+    case $1 in
+      -h|--help)
+        echo "Run Linkerd integration tests."
+        echo ""
+        echo "Optionally specify one of the following tests: [${test_names[*]}]"
+        echo ""
+        echo "Usage:"
+        echo "    ${0##*/} [--images] [--images-host ssh://linkerd-docker] [--name test-name] [--skip-kind-create] /path/to/linkerd"
+        echo ""
+        echo "Examples:"
+        echo "    # Run all tests in isolated clusters"
+        echo "    ${0##*/} /path/to/linkerd"
+        echo ""
+        echo "    # Run single test in isolated clusters"
+        echo "    ${0##*/} --name test-name /path/to/linkerd"
+        echo ""
+        echo "    # Skip KinD cluster creation and run all tests in default cluster context"
+        echo "    ${0##*/} --skip-kind-create /path/to/linkerd"
+        echo ""
+        echo "    # Load images from tar files located under the 'image-archives' directory"
+        echo "    # Note: This is primarly for CI"
+        echo "    ${0##*/} --images /path/to/linkerd"
+        echo ""
+        echo "    # Retrieve images from a remote docker instance and then load them into KinD"
+        echo "    # Note: This is primarly for CI"
+        echo "    ${0##*/} --images --images-host ssh://linkerd-docker /path/to/linkerd"
+        echo ""
+        echo "Available Commands:"
+        echo "    --name: the argument to this option is the specific test to run"
+        echo "    --skip-kind-create: skip KinD cluster creation step and run tests in an existing cluster."
+        echo "    --images: (Primarily for CI) use 'kind load image-archive' to load the images from local .tar files in the current directory."
+        echo "    --images-host: (Primarily for CI) the argument to this option is used as the remote docker instance from which images are first retrieved (using 'docker save') to be then loaded into KinD. This command requires --images."
+        exit 0
```

Comment:
The shellscripting folks tend to suggest using one echo command instead of multiple ones. No that a modern Threadripper CPU would stagger, but it is a relatively big performance difference between one echo command and multiple ones. Hence it would be something like

```
echo "Run Linkerd integration tests.

Optionally specify one of the following tests: [${test_names[*]}]

Usage:
    ${0##*/} [--images] [--images-host ssh://linkerd-docker] [--name test-name] [--skip-kind-create] /path/to/linkerd
[etc.]
"
```
which is also easier to read and maintain.